### PR TITLE
Parser instanceof

### PR DESCRIPTION
**Pull request by mickaelistria**
## What it does

This PR removes a few "noisy" lines and error-prone casts in Parser by leveraging instanceof pattern

## How to test

Just no regression should happen

## Author checklist

- [ ] I have thoroughly tested my changes (I leave it to CI)
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by jukzi** [comment]
is there a cleanup job for that refactoring, or manual work?

**Comment by mickaelistria** [comment]
manual work.

**Comment by mickaelistria** [comment]
cc @rgrunber @jjohnstn can you please review, and merge if OK?

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java [hunk 1]
```diff
@@ -2186,26 +2166,25 @@ protected void consumeBinaryExpressionWithName(int op) {
...
-					if (expr2 instanceof CharLiteral) { // string+char
+					if (expr2 instanceof CharLiteral char2) { // string+char
 						this.expressionStack[this.expressionPtr] =
-							((StringLiteral) expr1).extendWith((CharLiteral) expr2);
-					} else if (expr2 instanceof StringLiteral) { //string+string
+							string1.extendWith(char2);
+					} else if (expr2 instanceof StringLiteral string2) { //string+string
 						this.expressionStack[this.expressionPtr] =
-							((StringLiteral) expr1).extendWith((StringLiteral) expr2);
+							string2.extendWith(string2);
```
Comment (rgrunber) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java:2176]: I think

```suggestion
							string1.extendWith(string2);
```

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java [hunk 2]
```diff
@@ -1398,34 +1384,29 @@ protected ParameterizedQualifiedTypeReference computeQualifiedGenericsFromRightS
 		tokensSize ++;
 	} else if (rightSide instanceof SingleTypeReference) {
 		tokensSize ++;
-	} else if (rightSide instanceof ParameterizedQualifiedTypeReference) {
-		tokensSize += ((QualifiedTypeReference) rightSide).tokens.length;
```
Comment (rgrunber) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java:1402]: Why is this removed ? I doesn't appear any of the above `instanceof` checks would render it useless.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java [hunk 3]
```diff
@@ -1092,31 +1092,28 @@ public RecoveredElement buildInitialRecoveryState(){
 		this.endPosition = 0;
 		this.endStatementPosition = 0;
 		return element;
+	} else if (this.referenceContext instanceof AbstractMethodDeclaration methodDeclaration){
+		element = new RecoveredMethod(methodDeclaration, null, 0, this);
+		this.lastCheckPoint = ((AbstractMethodDeclaration) this.referenceContext).bodyStart;
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java:1097]: this line can also use methodDeclaration

**Review by jjohnstn — CHANGES_REQUESTED** [review]

**Comment by jjohnstn** [comment]
> is there a cleanup job for that refactoring, or manual work?

Just for clarification.  There is a cleanup for such refactoring on the Java Features tab for Java 16.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java (hunk 2 above)
Comment (mickaelistria) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java:1402]: The case of `ParameterizedQualifiedTypeReference` is already handled in the very exact same way in the next `instanceof QualifiedTypeReference`, so those like are actually useless. But as they're not related to instanceof, I can keep them and create another commit for it if you think it makes things better or easier.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java (hunk 1 above)
Comment (mickaelistria) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java]: indeed, thanks!

**Comment by iloveeclipse** [comment]
Mickael, 
1) looks like you pushed wrong branch.
2) It would be nice if you could just apply cleanup without changing the code otherwise.

**Comment by mickaelistria** [comment]
> looks like you pushed wrong branch.

Yes, it was a wrong push, but I fixed it in the meantime.

> It would be nice if you could just apply cleanup without changing the code otherwise.

I'm not sure I fully understand this sentence as all the change here is a cleanup. What is it concretely in this proposal you call "changing the code otherwise"?

**Comment by iloveeclipse** [comment]
I meant *automatic* cleanup.

**Comment by mickaelistria** [comment]
I didn't use an automatic cleanup for this change.

**Comment by iloveeclipse** [comment]
Why not?

**Comment by mickaelistria** [comment]
> Why not?

I wanted to learn about this class and used such manual work to familiarize myself with it. I also use the Generic Editor which doesn't have such trivial integration for cleanups, and I also even didn't know such clean up exists in an automated way.
Thinking further about the topic of cleanup, I think we could try to automate them in Platform/JDT/PDE, by having a bot running them and submitting PRs. I may spend some time on it in the future.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java [hunk 4]
```diff
@@ -1181,37 +1174,32 @@ public RecoveredElement buildInitialRecoveryState(){
...
+			if(node instanceof Block block) {
 				element = element.add(block, 0);
 				this.lastCheckPoint = block.sourceEnd + 1;
-			} else if(node instanceof LocalDeclaration) {
-				LocalDeclaration statement = (LocalDeclaration) node;
+			} else if(node instanceof LocalDeclaration statement) {
 				element = element.add(statement, 0);
 				this.lastCheckPoint = statement.sourceEnd + 1;
-			} else if(node instanceof Expression &&  ((Expression) node).isTrulyExpression()) {
+			} else if(node instanceof Expression statement &&  statement.isTrulyExpression()) {
```
Comment (szarnekow) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java:1188]: I think it should not be called statement if it’s an expression

Comment (szarnekow) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java:1188]: Though that was already the case prior to this cleanup :(

Comment (mickaelistria) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java:1188]: I wouldn't mind changing it by the way but then it would be "changing the code otherwise" as @iloveeclipse mentioned should be avoided.
I would recommend that you open a PR that change the name and if it is merged earlier, I'll adapt this patch.

**Review by rgrunber — APPROVED** [review]
I think all the issues have been addressed. Any additional style issues (statement used to refer to an expression) we could probably fix in a separate PR.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java [hunk 5]
```diff
@@ -2054,21 +2034,21 @@ protected void consumeBinaryExpression(int op) {
...
 						new BinaryExpression(expr1, expr2, PLUS);
 				}
-			} else if (expr1 instanceof StringLiteral) {
+			} else if (expr1 instanceof StringLiteral string) {
 				if (expr2 instanceof StringLiteral
 						&& ((expr1.bits & ASTNode.ParenthesizedMASK) >> ASTNode.ParenthesizedSHIFT) == 0) {
 					// string + string
 					this.expressionStack[this.expressionPtr] =
-						((StringLiteral) expr1).extendsWith((StringLiteral) expr2);
+						string.extendsWith((StringLiteral) expr2);
```
Comment (jukzi) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java:2042]: nitpick: this typcast could also be done with the instanceof in the if in front
